modulus/models/afno/modafno.py

```diff
@@ -0,0 +1,649 @@
+        Drop out rate in MLP, by default 0.0
+    activation_fn: nn.Module, optional
+        Activation function used in MLP, by default nn.GELU
+    norm_layer : nn.Module, optional
+        Normalization function, by default nn.LayerNorm
+    double_skip : bool, optional
+        Residual, by default True
+    sparsity_threshold : float, optional
+        Sparsity threshold (softshrink) of spectral features, by default 0.01
+    hard_thresholding_fraction : float, optional
+        Threshold for limiting number of modes used [0,1], by default 1
+    modulate_filter: bool, optional
+        Whether to compute the modulation for the FFT filter
+    modulate_mlp: bool, optional
+        Whether to compute the modulation for the MLP
+    scale_shift_mode: str
+        If 'complex' (default), compute the scale-shift operation using complex
+        operations. If 'real', use real operations.
+    """
+
+    def __init__(
+        self,
+        embed_dim: int,
+        mod_dim: int,
+        num_blocks: int = 8,
+        mlp_ratio: float = 4.0,
+        drop: float = 0.0,
+        activation_fn: nn.Module = nn.GELU(),
+        norm_layer: nn.Module = nn.LayerNorm,
+        double_skip: bool = True,
+        sparsity_threshold: float = 0.01,
+        hard_thresholding_fraction: float = 1.0,
+        modulate_filter: bool = True,
+        modulate_mlp: bool = True,
+        scale_shift_mode: str = "real",
+    ):
+        super().__init__()
+        self.norm1 = norm_layer(embed_dim)
+        if modulate_filter:
+            self.filter = ModAFNO2DLayer(
+                embed_dim,
+                mod_dim,
+                num_blocks,
+                sparsity_threshold,
+                hard_thresholding_fraction,
+                scale_shift_mode=scale_shift_mode,
+            )
+        else:
+            self.filter = AFNO2DLayer(
+                embed_dim, num_blocks, sparsity_threshold, hard_thresholding_fraction
+            )
+
+        self.norm2 = norm_layer(embed_dim)
+        mlp_latent_dim = int(embed_dim * mlp_ratio)
+        if modulate_mlp:
+            self.mlp = ModAFNOMlp(
+                in_features=embed_dim,
+                latent_features=mlp_latent_dim,
+                out_features=embed_dim,
+                mod_features=mod_dim,
+                activation_fn=activation_fn,
+                drop=drop,
+            )
+        else:
+            self.mlp = AFNOMlp(
+                in_features=embed_dim,
+                latent_features=mlp_latent_dim,
+                out_features=embed_dim,
+                activation_fn=activation_fn,
+                drop=drop,
+            )
+        self.double_skip = double_skip
+        self.modulate_filter = modulate_filter
+        self.modulate_mlp = modulate_mlp
+
+    def forward(self, x: Tensor, mod_embed: Tensor) -> Tensor:
+        residual = x
+        x = self.norm1(x)
+        x = self.filter(x, mod_embed) if self.modulate_filter else self.filter(x)
```

Comment:
Could so some thing like

filder_model = AFNOMlp()
self.filer = lambda x, mode_embed : filder_model(x)

to avoid having an if statement in your forward.

Reply:
Changed in https://github.com/NVIDIA/modulus/pull/547/commits/447c62e650e79dfc0e4342a93b1bd7780067661d, I implemented this a bit differently to avoid overwriting `self.filter` so existing checkpoints will work. Maybe check if this makes sense to you?